File: libparabolic/src/models/videoresolution.cpp
#include "models/videoresolution.h"
#include <limits>
#include <libnick/helpers/stringhelpers.h>
#include <libnick/localization/gettext.h>

using namespace Nickvision::Helpers;

namespace Nickvision::TubeConverter::Shared::Models
{
    VideoResolution::VideoResolution()
        : m_width{ std::numeric_limits<int>::max() },
        m_height{ std::numeric_limits<int>::max() }
    {

    }

    VideoResolution::VideoResolution(int width, int height)
        : m_width{ width },
        m_height{ height }
    {

    }

    VideoResolution::VideoResolution(boost::json::object json)
        : m_width{ json["Width"].is_int64() ? static_cast<int>(json["Width"].as_int64()) : std::numeric_limits<int>::max() },
        m_height{ json["Height"].is_int64() ? static_cast<int>(json["Height"].as_int64()) : std::numeric_limits<int>::max() }
    {

    }

    std::optional<VideoResolution> VideoResolution::parse(const std::string& value)
    {
        if(value == "Best" || value == _("Best"))
        {
            return VideoResolution{};
        }
        const std::vector<std::string> parts{ StringHelpers::split(value, "x") };
        if(parts.size() == 2)
        {
            try
            {
                int width{ std::stoi(parts[0]) };
                int height{ std::stoi(parts[1]) };
                return VideoResolution{ width, height };
            }
            catch(...) { }
        }
        return std::nullopt;
    }

    bool VideoResolution::isValid() const
    {
        return m_width > 0 && m_height > 0;
    }

    bool VideoResolution::isBest() const
    {
        return m_width == std::numeric_limits<int>::max() && m_height == std::numeric_limits<int>::max();
    }

    int VideoResolution::getWidth() const
    {
        return m_width;
    }

    int VideoResolution::getHeight() const
    {
        return m_height;
    }

    std::string VideoResolution::str() const
    {
        if(isBest())
        {
            return _("Best");
        }
        return std::to_string(m_width) + "x" + std::to_string(m_height);
    }

    boost::json::object VideoResolution::toJson() const
    {
        boost::json::object json;
        json["Width"] = m_width;
        json["Height"] = m_height;
        return json;
    }

    bool VideoResolution::operator==(const VideoResolution& other) const
    {
        return m_width == other.m_width && m_height == other.m_height;
    }

    bool VideoResolution::operator!=(const VideoResolution& other) const
    {
        return !operator==(other);
    }

    bool VideoResolution::operator<(const VideoResolution& other) const
    {
        return m_width < other.m_width || (m_width == other.m_width && m_height < other.m_height);
    }

    bool VideoResolution::operator>(const VideoResolution& other) const
    {
        return m_width > other.m_width || (m_width == other.m_width && m_height > other.m_height);
    }

    VideoResolution::operator bool() const
    {
        return isValid();
    }
}

File: Nickvision.Parabolic.Shared/Models/VideoResolution.cs
using Nickvision.Desktop.Globalization;
using System;

namespace Nickvision.Parabolic.Shared.Models;

public class VideoResolution : IComparable<VideoResolution>, IEquatable<VideoResolution>
{
    public static VideoResolution Best { get; }
    public static VideoResolution Worst { get; }

    public int Width { get; }
    public int Height { get; }

    static VideoResolution()
    {
        Best = new VideoResolution(int.MaxValue, int.MaxValue);
        Worst = new VideoResolution(-1, -1);
    }

    public VideoResolution(int width, int height)
    {
        Width = width;
        Height = height;
    }

    public static VideoResolution? Parse(string value, ITranslationService translator)
    {
        if (value == "Best" || value == translator._("Best"))
        {
            return Best;
        }
        else if (value == "Worst" || value == translator._("Worst"))
        {
            return Worst;
        }
        var parts = value.Split('x');
        if (parts.Length != 2)
        {
            return null;
        }
        if (int.TryParse(parts[0], out int width) && int.TryParse(parts[1], out int height))
        {
            return new VideoResolution(width, height);
        }
        return null;
    }

    public int CompareTo(VideoResolution? other) => other is null ? 1 : (Width * Height).CompareTo(other.Width * other.Height);

    public override bool Equals(object? obj) => obj is VideoResolution other && Equals(other);

    public bool Equals(VideoResolution? other) => other is not null && Width == other.Width && Height == other.Height;

    public override int GetHashCode() => HashCode.Combine(Width, Height);

    public override string ToString() => ToString(null);

    public string ToString(ITranslationService? translator)
    {
        if(Width == int.MaxValue && Height == int.MaxValue)
        {
            return translator?._("Best") ?? "Best";
        }
        else if(Width == -1 && Height == -1)
        {
            return translator?._("Worst") ?? "Worst";
        }
        return $"{Width}x{Height}";
    }

    public static bool operator >(VideoResolution left, VideoResolution right) => left.CompareTo(right) > 0;

    public static bool operator <(VideoResolution left, VideoResolution right) => left.CompareTo(right) < 0;

    public static bool operator >=(VideoResolution left, VideoResolution right) => left.CompareTo(right) >= 0;

    public static bool operator <=(VideoResolution left, VideoResolution right) => left.CompareTo(right) <= 0;

    public static bool operator ==(VideoResolution left, VideoResolution right) => left.Equals(right);

    public static bool operator !=(VideoResolution left, VideoResolution right) => !left.Equals(right);
}
